File: algorithms/18_ocaml/dp/lcs.ml
let lcs s1 s2 =
  let m = String.length s1 in
  let n = String.length s2 in
  let dp = Array.make_matrix (m + 1) (n + 1) 0 in
  for i = 1 to m do
    for j = 1 to n do
      if s1.[i - 1] = s2.[j - 1] then
        dp.(i).(j) <- dp.(i - 1).(j - 1) + 1
      else
        dp.(i).(j) <- max dp.(i - 1).(j) dp.(i).(j - 1)
    done
  done;
  dp.(m).(n)

File: algorithms/17_matlab/dp/lcs.m
function len = lcs(s1, s2)
% LCS Computes length of Longest Common Subsequence
    m = length(s1);
    n = length(s2);
    dp = zeros(m + 1, n + 1);
    for i = 1:m
        for j = 1:n
            if s1(i) == s2(j)
                dp(i + 1, j + 1) = dp(i, j) + 1;
            else
                dp(i + 1, j + 1) = max(dp(i + 1, j), dp(i, j + 1));
            end
        end
    end
    len = dp(m + 1, n + 1);
end
